File: solve/haskell/Select.hs
import Common 

select :: [a] -> [(a, [a])]
select [x] = [(x, [])]
select (x:xs) = (x, xs) : [(y, x:ys) | (y, ys) <- select xs]

permute :: [a] -> [[a]]
permute [x] = [[x]]
permute xs = [y:zs | (y, ys) <- select xs, zs <- permute ys]

main = do skeleton "unordered" permute

File: solve/lua/select.lua
#! /usr/bin/lua

require "common"

function permute(k)
	for i = 2, n do
		swap(i, k % i + 1)
		k = math.floor(k / i)
	end
end

skeleton(UNORDERED, function ()
	local top = 1
	for i = 2, n do
		top = top * i
	end

	for k = 0, top - 1 do
        undo()
		permute(k)
		output()
	end
end)
